File: server/application-server/src/main/java/de/tum/in/www1/hephaestus/leaderboard/LeaderboardService.java
package de.tum.in.www1.hephaestus.leaderboard;

import static de.tum.in.www1.hephaestus.shared.LeaguePointsConstants.POINTS_DEFAULT;
import static java.util.function.Function.identity;

import de.tum.in.www1.hephaestus.core.exception.EntityNotFoundException;
import de.tum.in.www1.hephaestus.gitprovider.label.Label;
import de.tum.in.www1.hephaestus.gitprovider.pullrequest.PullRequest;
import de.tum.in.www1.hephaestus.gitprovider.pullrequest.PullRequestInfoDTO;
import de.tum.in.www1.hephaestus.gitprovider.pullrequestreview.PullRequestReview;
import de.tum.in.www1.hephaestus.gitprovider.team.Team;
import de.tum.in.www1.hephaestus.gitprovider.team.TeamInfoDTO;
import de.tum.in.www1.hephaestus.gitprovider.team.TeamRepository;
import de.tum.in.www1.hephaestus.gitprovider.user.User;
import de.tum.in.www1.hephaestus.gitprovider.user.UserInfoDTO;
import de.tum.in.www1.hephaestus.gitprovider.user.UserRepository;
import de.tum.in.www1.hephaestus.workspace.Workspace;
import de.tum.in.www1.hephaestus.workspace.WorkspaceMembershipService;
import de.tum.in.www1.hephaestus.workspace.settings.WorkspaceTeamSettingsService;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Service for building leaderboard data.
 *
 * <p>XP scores are read from the activity event ledger, NOT recalculated on-the-fly.
 * This ensures consistency with the persisted XP values.
 */
@Service
public class LeaderboardService {

    private static final Logger log = LoggerFactory.getLogger(LeaderboardService.class);

    private final UserRepository userRepository;
    private final LeaderboardReviewQueryRepository leaderboardReviewQueryRepository;
    private final LeaderboardXpQueryService leaderboardXpQueryService;
    private final TeamRepository teamRepository;
    private final TeamPathResolver teamPathResolver;
    private final LeaguePointsService leaguePointsService;
    private final WorkspaceMembershipService workspaceMembershipService;
    private final WorkspaceTeamSettingsService workspaceTeamSettingsService;

    public LeaderboardService(
        UserRepository userRepository,
        LeaderboardReviewQueryRepository leaderboardReviewQueryRepository,
        LeaderboardXpQueryService leaderboardXpQueryService,
        TeamRepository teamRepository,
        TeamPathResolver teamPathResolver,
        LeaguePointsService leaguePointsService,
        WorkspaceMembershipService workspaceMembershipService,
        WorkspaceTeamSettingsService workspaceTeamSettingsService
    ) {
        this.userRepository = userRepository;
        this.leaderboardReviewQueryRepository = leaderboardReviewQueryRepository;
        this.leaderboardXpQueryService = leaderboardXpQueryService;
        this.teamRepository = teamRepository;
        this.teamPathResolver = teamPathResolver;
        this.leaguePointsService = leaguePointsService;
        this.workspaceMembershipService = workspaceMembershipService;
        this.workspaceTeamSettingsService = workspaceTeamSettingsService;
    }

    @Transactional(readOnly = true)
    public List<LeaderboardEntryDTO> createLeaderboard(
        Workspace workspace,
        Instant after,
        Instant before,
        String team,
        LeaderboardSortType sort,
        LeaderboardMode mode
    ) {
        if (mode == LeaderboardMode.INDIVIDUAL) {
            Optional<Team> resolvedTeam = "all".equals(team)
                ? Optional.empty()
                : teamPathResolver.resolveByPath(workspace, team);
            return createIndividualLeaderboard(workspace, after, before, resolvedTeam, sort);
        }

        // Team mode aggregates across all visible teams; the team filter is ignored on purpose (documented in OpenAPI).
        return createTeamLeaderboard(workspace, after, before, sort);
    }

    private List<LeaderboardEntryDTO> createIndividualLeaderboard(
        Workspace workspace,
        Instant after,
        Instant before,
        Optional<Team> team,
        LeaderboardSortType sort
    ) {
        return createIndividualLeaderboard(workspace, after, before, team, sort, null);
    }

    private List<LeaderboardEntryDTO> createIndividualLeaderboard(
        Workspace workspace,
        Instant after,
        Instant before,
        Optional<Team> team,
        LeaderboardSortType sort,
        Map<Long, List<Team>> teamHierarchy
    ) {
        if (workspace == null || workspace.getId() == null) {
            log.warn("Skipping leaderboard dataset creation because workspace id is missing.");
            return Collections.emptyList();
        }

        log.info(
            "Creating leaderboard dataset with timeframe: {} - {} and team: {} for workspace {}",
            after,
            before,
            team.isEmpty() ? "all" : team.get().getName(),
            workspace.getWorkspaceSlug()
        );

        Long workspaceId = workspace.getId();

        // Collect team IDs for filtering (if team specified)
        Set<Long> teamIds = team
            .map(t ->
                teamPathResolver.collectDescendantIds(
                    t,
                    Objects.requireNonNullElseGet(teamHierarchy, () -> teamPathResolver.buildHierarchy(workspace))
                )
            )
            .orElse(Collections.emptySet());

        // ========================================================================
        // XP and breakdown from activity events (source of truth)
        // ========================================================================
        Map<Long, LeaderboardUserXp> activityData = new HashMap<>(
            leaderboardXpQueryService.getLeaderboardData(workspaceId, after, before, teamIds)
        );

        // ========================================================================
        // Include ALL team members, even those with zero activity
        // ========================================================================
        List<User> allTeamMembers;
        if (team.isPresent() && !teamIds.isEmpty()) {
            allTeamMembers = userRepository.findAllByTeamIds(teamIds);
        } else if (workspace.getAccountLogin() != null) {
            allTeamMembers = userRepository.findAllHumanInTeamsOfOrganization(workspace.getAccountLogin());
        } else {
            allTeamMembers = Collections.emptyList();
        }

        // Add zero-score entries for team members without activity
        for (User member : allTeamMembers) {
            if (member.getId() != null && !activityData.containsKey(member.getId())) {
                // New record signature: user, totalScore, eventCount, approvals, changeRequests,
                // comments, unknowns, issueComments, codeComments, reviewedPrCount
                activityData.put(member.getId(), new LeaderboardUserXp(member, 0, 0, 0, 0, 0, 0, 0, 0, 0));
            }
        }

        if (activityData.isEmpty()) {
            log.info("No team members found for leaderboard");
            return Collections.emptyList();
        }

        // ========================================================================
        // Fetch PR data for display (not for scoring)
        // ========================================================================
        Set<Long> actorIds = activityData.keySet();

        List<PullRequestReview> reviews;
        if (team.isPresent() && !teamIds.isEmpty()) {
            reviews = leaderboardReviewQueryRepository.findAllInTimeframeOfTeams(after, before, teamIds, workspaceId);
        } else {
            reviews = leaderboardReviewQueryRepository.findAllInTimeframe(after, before, workspaceId);
        }

        Map<Long, List<PullRequestReview>> reviewsByUserId = reviews
            .stream()
            .filter(r -> r.getAuthor() != null && r.getAuthor().getId() != null)
            .filter(r -> actorIds.contains(r.getAuthor().getId()))
            .collect(Collectors.groupingBy(r -> r.getAuthor().getId()));

        // ========================================================================
        // Get league points for ranking
        // ========================================================================
        Collection<User> users = activityData.values().stream().map(LeaderboardUserXp::user).toList();

        Map<Long, Integer> leaguePointsByUserId = workspaceMembershipService.getLeaguePointsSnapshot(
            users,
            workspaceId
        );

        // ========================================================================
        // Sort by specified criteria
        // ========================================================================
        Comparator<Map.Entry<Long, LeaderboardUserXp>> comparator = comparatorForActivityData(
            sort,
            leaguePointsByUserId
        );

        List<Long> ranking = activityData.entrySet().stream().sorted(comparator).map(Map.Entry::getKey).toList();

        // ========================================================================
        // Build DTOs
        // ========================================================================
        List<LeaderboardEntryDTO> result = new ArrayList<>();
        for (int index = 0; index < ranking.size(); index++) {
            Long userId = ranking.get(index);
            LeaderboardUserXp data = activityData.get(userId);
            if (data == null) {
                continue;
            }

            User user = data.user();
            int score = data.totalScore();

            UserInfoDTO userDto = UserInfoDTO.fromUser(user, leaguePointsByUserId.getOrDefault(userId, POINTS_DEFAULT));

            List<PullRequestReview> userReviews = reviewsByUserId.getOrDefault(userId, Collections.emptyList());

            // Extract reviewed PRs for display popover
            List<PullRequestInfoDTO> reviewedPullRequests = userReviews
                .stream()
                .map(PullRequestReview::getPullRequest)
                .filter(Objects::nonNull)
                .filter(pr -> pr.getAuthor() == null || !Objects.equals(pr.getAuthor().getId(), userId))
                .collect(Collectors.toMap(PullRequest::getId, pr -> pr, (a, b) -> a))
                .values()
                .stream()
                .map(PullRequestInfoDTO::fromPullRequest)
                .collect(Collectors.toList());

            // Stats come from activity events (pre-computed, not recalculated)
            LeaderboardEntryDTO entry = new LeaderboardEntryDTO(
                index + 1,
                score,
                userDto,
                null,
                reviewedPullRequests,
                data.reviewedPullRequestCount(),
                data.approvals(),
                data.changeRequests(),
                data.comments() + data.issueComments(),
                data.unknowns(), // numberOfUnknowns - now properly tracked in activity events
                data.codeComments()
            );
            result.add(entry);
        }

        return result;
    }

    /**
     * Comparator for activity-based leaderboard data.
     */
    private Comparator<Map.Entry<Long, LeaderboardUserXp>> comparatorForActivityData(
        LeaderboardSortType sort,
        Map<Long, Integer> leaguePointsByUserId
    ) {
        return (a, b) -> {
            LeaderboardUserXp dataA = a.getValue();
            LeaderboardUserXp dataB = b.getValue();

            return switch (sort) {
                case SCORE -> Integer.compare(dataB.totalScore(), dataA.totalScore());
                case LEAGUE_POINTS -> {
                    int lpA = leaguePointsByUserId.getOrDefault(a.getKey(), POINTS_DEFAULT);
                    int lpB = leaguePointsByUserId.getOrDefault(b.getKey(), POINTS_DEFAULT);
                    // Tie-breaker: use score if league points are equal
                    int cmp = Integer.compare(lpB, lpA);
                    if (cmp != 0) {
                        yield cmp;
                    }
                    yield Integer.compare(dataB.totalScore(), dataA.totalScore());
                }
            };
        };
    }

    private List<LeaderboardEntryDTO> createTeamLeaderboard(
        Workspace workspace,
        Instant after,
        Instant before,
        LeaderboardSortType sort
    ) {
        log.info(
            "Creating team leaderboard dataset with timeframe: {} - {} in workspace {}",
            after,
            before,
            workspace.getWorkspaceSlug()
        );

        Map<Long, List<Team>> teamHierarchy = teamPathResolver.buildHierarchy(workspace);
        // Use workspace-scoped hidden settings instead of deprecated Team.hidden field
        Set<Long> hiddenTeamIds = workspaceTeamSettingsService.getHiddenTeamIds(workspace.getId());
        List<Team> targetTeams = workspace.getAccountLogin() == null
            ? List.of()
            : teamRepository
                  .findAllByOrganizationIgnoreCase(workspace.getAccountLogin())
                  .stream()
                  .filter(team -> team.getId() != null && !hiddenTeamIds.contains(team.getId()))
                  .toList();

        if (targetTeams.isEmpty()) {
            log.info("No teams found for team leaderboard in workspace {}", workspace.getWorkspaceSlug());
            return Collections.emptyList();
        }

        Map<Team, TeamStats> teamStatsById = targetTeams
            .stream()
            .collect(
                Collectors.toMap(identity(), teamEntity -> {
                    List<LeaderboardEntryDTO> entries = createIndividualLeaderboard(
                        workspace,
                        after,
                        before,
                        Optional.of(teamEntity),
                        LeaderboardSortType.SCORE,
                        teamHierarchy
                    );
                    return aggregateTeamStats(entries);
                })
            );

        List<Map.Entry<Team, TeamStats>> sorted = teamStatsById
            .entrySet()
            .stream()
            .sorted((e1, e2) -> {
                int cmp;
                if (sort == LeaderboardSortType.SCORE) {
                    cmp = Integer.compare(e2.getValue().score(), e1.getValue().score());
                } else {
                    cmp = Integer.compare(e2.getValue().leaguePoints(), e1.getValue().leaguePoints());
                }
                if (cmp != 0) {
                    return cmp;
                }
                return e1.getKey().getName().compareTo(e2.getKey().getName());
            })
            .toList();

        List<LeaderboardEntryDTO> result = new ArrayList<>();
        Long workspaceId = workspace.getId();
        for (int i = 0; i < sorted.size(); i++) {
            Team teamEntity = sorted.get(i).getKey();
            TeamStats stats = sorted.get(i).getValue();
            int score = sort == LeaderboardSortType.SCORE ? stats.score() : stats.leaguePoints();

            // Get workspace-scoped settings for this team
            boolean isHiddenInWorkspace = hiddenTeamIds.contains(teamEntity.getId());
            Set<Label> workspaceLabels = workspaceTeamSettingsService.getTeamLabelFilters(
                workspaceId,
                teamEntity.getId()
            );
            Set<Long> hiddenRepoIds = workspaceTeamSettingsService.getHiddenRepositoryIdsByTeams(
                workspaceId,
                Set.of(teamEntity.getId())
            );

            LeaderboardEntryDTO entry = new LeaderboardEntryDTO(
                i + 1,
                score,
                null,
                TeamInfoDTO.fromTeamWithWorkspaceSettings(
                    teamEntity,
                    isHiddenInWorkspace,
                    workspaceLabels,
                    hiddenRepoIds
                ),
                stats.reviewedPullRequests(),
                stats.numberOfReviewedPRs(),
                stats.numberOfApprovals(),
                stats.numberOfChangeRequests(),
                stats.numberOfComments(),
                stats.numberOfUnknowns(),
                stats.numberOfCodeComments()
            );
            result.add(entry);
        }
        return result;
    }

    @Transactional(readOnly = true)
    public LeagueChangeDTO getUserLeagueStats(Workspace workspace, String login, LeaderboardEntryDTO entry) {
        User user = userRepository.findByLogin(login).orElseThrow(() -> new EntityNotFoundException("User", login));

        if (workspace == null || workspace.getId() == null) {
            throw new IllegalStateException("Workspace context is required to compute league stats");
        }

        Long workspaceId = workspace.getId();

        int currentLeaguePoints = workspaceMembershipService.getCurrentLeaguePoints(workspaceId, user);
        int projectedNewPoints = leaguePointsService.calculateNewPoints(user, currentLeaguePoints, entry);
        return new LeagueChangeDTO(user.getLogin(), projectedNewPoints - currentLeaguePoints);
    }

    private TeamStats aggregateTeamStats(List<LeaderboardEntryDTO> entries) {
        List<PullRequestInfoDTO> reviewedPullRequests = new ArrayList<>(
            entries
                .stream()
                .flatMap(e -> e.reviewedPullRequests().stream())
                .collect(Collectors.toMap(PullRequestInfoDTO::id, p -> p, (a, b) -> a))
                .values()
        );

        int score = entries.stream().mapToInt(LeaderboardEntryDTO::score).sum();
        int leaguePoints = entries
            .stream()
            .mapToInt(e -> e.user() == null ? 0 : e.user().leaguePoints())
            .sum();
        int numberOfReviewedPRs = entries.stream().mapToInt(LeaderboardEntryDTO::numberOfReviewedPRs).sum();
        int numberOfApprovals = entries.stream().mapToInt(LeaderboardEntryDTO::numberOfApprovals).sum();
        int numberOfChangeRequests = entries.stream().mapToInt(LeaderboardEntryDTO::numberOfChangeRequests).sum();
        int numberOfComments = entries.stream().mapToInt(LeaderboardEntryDTO::numberOfComments).sum();
        int numberOfUnknowns = entries.stream().mapToInt(LeaderboardEntryDTO::numberOfUnknowns).sum();
        int numberOfCodeComments = entries.stream().mapToInt(LeaderboardEntryDTO::numberOfCodeComments).sum();

        return new TeamStats(
            score,
            leaguePoints,
            reviewedPullRequests,
            numberOfReviewedPRs,
            numberOfApprovals,
            numberOfChangeRequests,
            numberOfComments,
            numberOfUnknowns,
            numberOfCodeComments
        );
    }

    private record TeamStats(
        int score,
        int leaguePoints,
        List<PullRequestInfoDTO> reviewedPullRequests,
        int numberOfReviewedPRs,
        int numberOfApprovals,
        int numberOfChangeRequests,
        int numberOfComments,
        int numberOfUnknowns,
        int numberOfCodeComments
    ) {}
}

File: server/application-server/src/main/java/de/tum/in/www1/hephaestus/leaderboard/LeaderboardService.kt
package de.tum.`in`.www1.hephaestus.leaderboard

import de.tum.`in`.www1.hephaestus.gitprovider.issuecomment.IssueComment
import de.tum.`in`.www1.hephaestus.gitprovider.issuecomment.IssueCommentRepository
import de.tum.`in`.www1.hephaestus.gitprovider.pullrequest.PullRequest
import de.tum.`in`.www1.hephaestus.gitprovider.pullrequest.PullRequestInfoDTO
import de.tum.`in`.www1.hephaestus.gitprovider.pullrequestreview.PullRequestReview
import de.tum.`in`.www1.hephaestus.gitprovider.pullrequestreview.PullRequestReviewRepository
import de.tum.`in`.www1.hephaestus.gitprovider.team.Team
import de.tum.`in`.www1.hephaestus.gitprovider.team.TeamInfoDTO
import de.tum.`in`.www1.hephaestus.gitprovider.team.TeamRepository
import de.tum.`in`.www1.hephaestus.gitprovider.user.User
import de.tum.`in`.www1.hephaestus.gitprovider.user.UserInfoDTO
import de.tum.`in`.www1.hephaestus.gitprovider.user.UserRepository
import jakarta.transaction.Transactional
import java.time.Instant
import java.util.Optional
import kotlin.collections.ArrayDeque
import kotlin.math.ceil
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class LeaderboardService(
    private val userRepository: UserRepository,
    private val pullRequestReviewRepository: PullRequestReviewRepository,
    private val issueCommentRepository: IssueCommentRepository,
    private val scoringService: ScoringService,
    private val teamRepository: TeamRepository,
    private val leaguePointsCalculationService: LeaguePointsCalculationService,
) {

    private val logger = LoggerFactory.getLogger("LeaderboardService")

    @Transactional
    fun createLeaderboard(
        after: Instant,
        before: Instant,
        team: String,
        sort: LeaderboardSortType,
        mode: LeaderboardMode,
    ): List<LeaderboardEntryDTO> {
        return when (mode) {
            LeaderboardMode.INDIVIDUAL -> createIndividualLeaderboard(after, before, if (team == "all") null else resolveTeamByPath(team), sort)
            LeaderboardMode.TEAM -> createTeamLeaderboard(after, before)
        }
    }

    fun createLeaderboard(
        after: Instant,
        before: Instant,
        team: Optional<String>,
        sort: Optional<LeaderboardSortType>,
        mode: Optional<LeaderboardMode>,
    ): List<LeaderboardEntryDTO> = createLeaderboard(
        after,
        before,
        team.orElse(null),
        sort.orElse(null),
        mode.orElse(null),
    )

    private fun createIndividualLeaderboard(
        after: Instant,
        before: Instant,
        team: Team?,
        sort: LeaderboardSortType,
        teamHierarchy: Map<Long?, List<Team>>? = null,
    ): List<LeaderboardEntryDTO> {
        logger.info(
            "Creating leaderboard dataset with timeframe: {} - {} and team: {}",
            after,
            before,
            team?.name ?: "all",
        )

        val reviews: List<PullRequestReview>
        val issueComments: List<IssueComment>

        val teamIds: Set<Long> = if (team != null) {
            val hierarchy = teamHierarchy ?: buildTeamHierarchy()
            collectTeamAndDescendantIds(team, hierarchy)
        } else {
            emptySet()
        }

        if (team != null) {
            reviews = pullRequestReviewRepository.findAllInTimeframeOfTeams(after, before, teamIds)
            issueComments = issueCommentRepository.findAllInTimeframeOfTeams(after, before, teamIds, true)
        } else {
            reviews = pullRequestReviewRepository.findAllInTimeframe(after, before)
            issueComments = issueCommentRepository.findAllInTimeframe(after, before, true)
        }

        val usersById: MutableMap<Long, User> = reviews
            .mapNotNull { it.author }
            .associateBy { it.id }
            .toMutableMap()

        issueComments.mapNotNull { it.author }.forEach { usersById.putIfAbsent(it.id, it) }

        if (team != null) {
            if (teamIds.isNotEmpty()) {
                userRepository.findAllByTeamIds(teamIds).forEach { usersById.putIfAbsent(it.id, it) }
            }
        } else {
            userRepository.findAllHumanInTeams().forEach { usersById.putIfAbsent(it.id, it) }
        }

        val reviewsByUserId: Map<Long, List<PullRequestReview>> = reviews
            .filter { it.author != null }
            .groupBy { it.author.id }
        val issueCommentsByUserId: Map<Long, List<IssueComment>> = issueComments
            .filter { it.author != null }
            .groupBy { it.author.id }

        val scoresByUserId: MutableMap<Long, Int> = reviewsByUserId
            .mapValues { (userId, userReviews) ->
                calculateTotalScore(userReviews, issueCommentsByUserId[userId].orEmpty())
            }
            .toMutableMap()

        usersById.keys.forEach { userId -> scoresByUserId.putIfAbsent(userId, 0) }

        val ranking: List<Long> = scoresByUserId
            .entries
            .sortedWith(comparatorFor(sort, usersById, reviewsByUserId, issueCommentsByUserId))
            .map { it.key }

        return ranking.mapIndexed { index, userId ->
            val score = scoresByUserId[userId] ?: 0
            val user = usersById[userId]?.let(UserInfoDTO::fromUser)
            val userReviews = reviewsByUserId[userId].orEmpty()
            val userIssueComments = issueCommentsByUserId[userId].orEmpty()

            val reviewedPullRequests = userReviews
                .mapNotNull { it.pullRequest }
                .filter { it.author?.id != userId }
                .associateBy(PullRequest::getId)
                .values
                .map(PullRequestInfoDTO::fromPullRequest)

            val numberOfReviewedPRs = userReviews.mapNotNull { it.pullRequest?.id }.toSet().size
            val numberOfApprovals = userReviews.count { it.state == PullRequestReview.State.APPROVED }
            val numberOfChangeRequests = userReviews.count { it.state == PullRequestReview.State.CHANGES_REQUESTED }
            val numberOfComments = userReviews.count { it.state == PullRequestReview.State.COMMENTED && it.body != null } + userIssueComments.size
            val numberOfUnknowns = userReviews.count { it.state == PullRequestReview.State.UNKNOWN && it.body != null }
            val numberOfCodeComments = userReviews.sumOf { it.comments.size }

            LeaderboardEntryDTO(
                rank = index + 1,
                score = score,
                user = user,
                team = null,
                reviewedPullRequests = reviewedPullRequests,
                numberOfReviewedPRs = numberOfReviewedPRs,
                numberOfApprovals = numberOfApprovals,
                numberOfChangeRequests = numberOfChangeRequests,
                numberOfComments = numberOfComments,
                numberOfUnknowns = numberOfUnknowns,
                numberOfCodeComments = numberOfCodeComments,
            )
        }
    }

    private fun createTeamLeaderboard(
        after: Instant,
        before: Instant,
    ): List<LeaderboardEntryDTO> {
        logger.info(
            "Creating team leaderboard dataset with timeframe: {} - {}",
            after,
            before,
        )

        val allTeams = teamRepository.findAll()
        val teamHierarchy = allTeams.groupBy { it.parentId }
        val targetTeams = allTeams.filter { !it.isHidden }

        if (targetTeams.isEmpty()) {
            logger.info("❌ No teams found for team leaderboard")
            return emptyList()
        }

        val teamStatsById = targetTeams.associateWith { teamEntity ->
            val entries = createIndividualLeaderboard(after, before, teamEntity, LeaderboardSortType.SCORE, teamHierarchy)
            aggregateTeamStats(entries)
        }

        return teamStatsById
            .entries
            .sortedWith(
                compareByDescending<Map.Entry<Team, TeamStats>> { it.value.score }
                    .thenBy { it.key.name }
            )
            .mapIndexed { index, (teamEntity, stats) ->
                LeaderboardEntryDTO(
                    rank = index + 1,
                    score = stats.score,
                    user = null,
                    team = TeamInfoDTO.fromTeam(teamEntity),
                    reviewedPullRequests = stats.reviewedPullRequests,
                    numberOfReviewedPRs = stats.numberOfReviewedPRs,
                    numberOfApprovals = stats.numberOfApprovals,
                    numberOfChangeRequests = stats.numberOfChangeRequests,
                    numberOfComments = stats.numberOfComments,
                    numberOfUnknowns = stats.numberOfUnknowns,
                    numberOfCodeComments = stats.numberOfCodeComments,
                )
            }
    }

    private fun comparatorFor(
        sortType: LeaderboardSortType,
        usersById: Map<Long, User>,
        reviewsByUserId: Map<Long, List<PullRequestReview>>,
        issueCommentsByUserId: Map<Long, List<IssueComment>>,
    ): Comparator<Map.Entry<Long, Int>> = when (sortType) {
        LeaderboardSortType.LEAGUE_POINTS -> compareByLeaguePoints(usersById, reviewsByUserId, issueCommentsByUserId)
        LeaderboardSortType.SCORE -> compareByScore(reviewsByUserId, issueCommentsByUserId)
    }

    private fun compareByLeaguePoints(
        usersById: Map<Long, User>,
        reviewsByUserId: Map<Long, List<PullRequestReview>>,
        issueCommentsByUserId: Map<Long, List<IssueComment>>,
    ): Comparator<Map.Entry<Long, Int>> = Comparator { e1, e2 ->
        val e1LeaguePoints = usersById[e1.key]?.leaguePoints ?: 0
        val e2LeaguePoints = usersById[e2.key]?.leaguePoints ?: 0
        val leagueCompare = e2LeaguePoints.compareTo(e1LeaguePoints)
        if (leagueCompare != 0) {
            leagueCompare
        } else {
            compareByScore(reviewsByUserId, issueCommentsByUserId).compare(e1, e2)
        }
    }

    private fun compareByScore(
        reviewsByUserId: Map<Long, List<PullRequestReview>>,
        issueCommentsByUserId: Map<Long, List<IssueComment>>,
    ): Comparator<Map.Entry<Long, Int>> = Comparator { e1, e2 ->
        val scoreCompare = e2.value.compareTo(e1.value)
        if (scoreCompare != 0) {
            scoreCompare
        } else {
            val e1ReviewComments = reviewsByUserId[e1.key].orEmpty().sumOf { it.comments.size }
            val e2ReviewComments = reviewsByUserId[e2.key].orEmpty().sumOf { it.comments.size }
            val e1IssueComments = issueCommentsByUserId[e1.key].orEmpty().size
            val e2IssueComments = issueCommentsByUserId[e2.key].orEmpty().size
            val e1TotalComments = e1ReviewComments + e1IssueComments
            val e2TotalComments = e2ReviewComments + e2IssueComments
            e2TotalComments.compareTo(e1TotalComments)
        }
    }

    private fun calculateTotalScore(
        reviews: List<PullRequestReview>,
        issueComments: List<IssueComment>,
    ): Int {
        val numberOfIssueComments = issueComments
            .filter { issueComment ->
                val issueAuthorId = issueComment.issue?.author?.id
                val commentAuthorId = issueComment.author?.id
                issueAuthorId != null && commentAuthorId != null && issueAuthorId != commentAuthorId
            }
            .size

        val reviewsByPullRequest = reviews.groupBy { it.pullRequest.id }

        val totalScore = reviewsByPullRequest
            .values
            .sumOf { pullRequestReviews ->
                scoringService.calculateReviewScore(pullRequestReviews, numberOfIssueComments)
            }

        return ceil(totalScore).toInt()
    }

    @Transactional
    fun getUserLeagueStats(login: String, entry: LeaderboardEntryDTO): LeagueChangeDTO {
        val user = userRepository.findByLogin(login).orElseThrow { IllegalArgumentException("User not found with login: $login") }
        val currentLeaguePoints = user.leaguePoints
        val projectedNewPoints = leaguePointsCalculationService.calculateNewPoints(user, entry)
        return LeagueChangeDTO(user.login, projectedNewPoints - currentLeaguePoints)
    }

    private fun resolveTeamByPath(path: String?): Team? {
        if (path.isNullOrBlank()) {
            return null
        }

        val parts = path.split(" / ")
        val leaf = parts.last()
        val candidates = teamRepository.findAllByName(leaf)
        if (candidates.isEmpty()) {
            return null
        }

        val cache = mutableMapOf<Long, Team>()
        var currentByCandidate = mutableMapOf<Long, Team>()

        candidates.forEach { candidate ->
            val candidateId = candidate.id ?: return@forEach
            cache[candidateId] = candidate
            currentByCandidate[candidateId] = candidate
        }

        if (currentByCandidate.isEmpty()) {
            return null
        }

        if (parts.size == 1) {
            val sole = currentByCandidate.keys.singleOrNull() ?: return null
            return cache[sole]
        }

        for (index in parts.size - 2 downTo 0) {
            if (currentByCandidate.size <= 1) {
                break
            }

            val expected = parts[index]
            val nextVisibleByCandidate = mutableMapOf<Long, Team?>()
            var pendingResolution = true

            while (pendingResolution) {
                pendingResolution = false
                val missingIds = mutableSetOf<Long>()

                currentByCandidate.forEach { (candidateId, cursor) ->
                    if (candidateId in nextVisibleByCandidate) {
                        return@forEach
                    }

                    var parentId = cursor.parentId
                    while (parentId != null) {
                        val parent = cache[parentId]
                        if (parent == null) {
                            missingIds.add(parentId)
                            break
                        }
                        if (!parent.isHidden) {
                            nextVisibleByCandidate[candidateId] = parent
                            break
                        }
                        parentId = parent.parentId
                    }

                    if (parentId == null && candidateId !in nextVisibleByCandidate) {
                        nextVisibleByCandidate[candidateId] = null
                    }
                }

                if (missingIds.isNotEmpty()) {
                    teamRepository.findAllById(missingIds).forEach { parent ->
                        val parentId = parent.id ?: return@forEach
                        cache.putIfAbsent(parentId, parent)
                    }
                    pendingResolution = true
                }
            }

            val filtered = mutableMapOf<Long, Team>()
            currentByCandidate.forEach { (candidateId, _) ->
                val nextVisible = nextVisibleByCandidate[candidateId]
                if (nextVisible != null && nextVisible.name == expected) {
                    filtered[candidateId] = nextVisible
                    nextVisible.id?.let { cache.putIfAbsent(it, nextVisible) }
                }
            }

            currentByCandidate = filtered

            if (currentByCandidate.isEmpty()) {
                return null
            }

            if (currentByCandidate.size == 1) {
                val onlyId = currentByCandidate.keys.first()
                return cache[onlyId]
            }
        }

        if (currentByCandidate.size > 1) {
            preloadAncestors(currentByCandidate.values, cache)
            for (candidateId in currentByCandidate.keys) {
                val candidate = cache[candidateId]
                if (candidate != null && equalsVisiblePath(candidate, parts, cache)) {
                    return candidate
                }
            }
            logger.warn(
                "Ambiguous team path '{}' resolved to multiple candidates; picking first.",
                sanitizeForLog(path),
            )
        }

        val anyId = currentByCandidate.keys.firstOrNull() ?: return null
        return cache[anyId]
    }

    private fun equalsVisiblePath(team: Team, parts: List<String>, cache: Map<Long, Team>): Boolean {
        var index = parts.size - 1
        var current: Team? = team
        while (current != null) {
            if (!current.isHidden) {
                if (index < 0 || parts[index] != current.name) {
                    return false
                }
                index -= 1
            }
            val parentId = current.parentId
            current = parentId?.let(cache::get)
        }
        return index < 0
    }

    private fun preloadAncestors(teams: Collection<Team>, cache: MutableMap<Long, Team>) {
        var pending = teams
            .mapNotNull { it.parentId }
            .filterNot { cache.containsKey(it) }
            .toMutableSet()

        while (pending.isNotEmpty()) {
            val nextRound = mutableSetOf<Long>()
            teamRepository.findAllById(pending).forEach { parent ->
                val parentId = parent.id ?: return@forEach
                if (!cache.containsKey(parentId)) {
                    cache[parentId] = parent
                }
                val ancestorId = parent.parentId
                if (ancestorId != null && !cache.containsKey(ancestorId)) {
                    nextRound.add(ancestorId)
                }
            }
            pending = nextRound
        }
    }

    private fun sanitizeForLog(input: String?): String? = input?.replace(Regex("[\r\n]"), "")

    private data class TeamStats(
        val score: Int,
        val reviewedPullRequests: List<PullRequestInfoDTO>,
        val numberOfReviewedPRs: Int,
        val numberOfApprovals: Int,
        val numberOfChangeRequests: Int,
        val numberOfComments: Int,
        val numberOfUnknowns: Int,
        val numberOfCodeComments: Int,
    )

    private fun aggregateTeamStats(entries: List<LeaderboardEntryDTO>): TeamStats {
        val reviewedPullRequests = entries
            .flatMap { it.reviewedPullRequests }
            .associateBy { it.id }
            .values
            .toList()

        return TeamStats(
            score = entries.sumOf { it.score },
            reviewedPullRequests = reviewedPullRequests,
            numberOfReviewedPRs = entries.sumOf { it.numberOfReviewedPRs },
            numberOfApprovals = entries.sumOf { it.numberOfApprovals },
            numberOfChangeRequests = entries.sumOf { it.numberOfChangeRequests },
            numberOfComments = entries.sumOf { it.numberOfComments },
            numberOfUnknowns = entries.sumOf { it.numberOfUnknowns },
            numberOfCodeComments = entries.sumOf { it.numberOfCodeComments },
        )
    }

    private fun buildTeamHierarchy(): Map<Long?, List<Team>> = teamRepository.findAll().groupBy { it.parentId }

    private fun collectTeamAndDescendantIds(team: Team, hierarchy: Map<Long?, List<Team>>): Set<Long> {
        val result = mutableSetOf<Long>()
        val queue = ArrayDeque<Team>()
        queue.add(team)

        while (queue.isNotEmpty()) {
            val current = queue.removeFirst()
            val currentId = current.id ?: continue
            if (result.add(currentId)) {
                hierarchy[currentId].orEmpty().forEach { child -> queue.add(child) }
            }
        }

        return result
    }
}
